src/data_provider/src/sysInfoMac.cpp

```diff
@@ -58,45 +61,188 @@ static const std::map<std::string, int> s_mapPackagesDirectories =
+        // what we get on kinfo.
+        struct proc_taskinfo pti;
+
+        if (sizeof(pti) != proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)))
+        {
+            throw std::runtime_error("proc_pidinfo: failed to get process info by pid");
+        }
+
+        virtualMemorySizeKiB = pti.pti_virtual_size / 1024;
     }
+    // Might return null if the process disappears while retrieving its data.
+    nlohmann::json toJson()
+    {
+        nlohmann::json result{};
+        const auto maxBufferSize { sysconf(_SC_GETPW_R_SIZE_MAX) };
 
-    jsProcessInfo["priority"]   = taskInfo.ptinfo.pti_priority;
-    jsProcessInfo["nice"]       = taskInfo.pbsd.pbi_nice;
-    jsProcessInfo["vm_size"]    = taskInfo.ptinfo.pti_virtual_size / KByte;
-    jsProcessInfo["start_time"] = taskInfo.pbsd.pbi_start_tvsec;
-    return jsProcessInfo;
-}
+        if (maxBufferSize == -1)
+        {
+            throw std::system_error
+            {
+                errno,
+                std::system_category(),
+                "Error getting max buffer size for passwd buffers."
+            };
+        }
+
+        std::vector<char> buff(maxBufferSize, '\0');
+        passwd userInfo;
+        group groupInfo;
+        // infoPtr will point to "info" if calls to getpwuid_r are successful.
+        passwd* userInfoPtr { nullptr };
+        group* groupInfoPtr { nullptr };
+
+        result["pid"]        = std::to_string(pid);
+        result["name"]       = std::move(name);
+        result["state"]      = s_mapTaskInfoState[state];
+        result["ppid"]       = ppid;
+        result["priority"]   = priority;
+        result["nice"]       = nice;
+        result["vm_size"]    = virtualMemorySizeKiB;
+        result["start_time"] = startTime;
+
+        memset(&userInfo, 0, sizeof userInfo);
+        getpwuid_r(euid, &userInfo, buff.data(), buff.size(), &userInfoPtr);
+
+        if (userInfoPtr)
+        {
+            if (userInfoPtr->pw_name)
+            {
+                result["euser"]  = userInfoPtr->pw_name;
+            }
+        }
+
+        memset(&userInfo, 0, sizeof userInfo);
+        getpwuid_r(ruid, &userInfo, buff.data(), buff.size(), &userInfoPtr);
+
+        if (userInfoPtr)
+        {
+            if (userInfoPtr->pw_name)
+            {
+                result["ruser"]  = userInfoPtr->pw_name;
+            }
+        }
+
+        memset(&groupInfo, 0, sizeof groupInfo);
+        getgrgid_r(rgid, &groupInfo, buff.data(), buff.size(), &groupInfoPtr);
+
+        if (groupInfoPtr)
+        {
+            if (groupInfoPtr->gr_name)
+            {
+                result["rgroup"] = groupInfoPtr->gr_name;
+            }
+        }
+
+        return result;
+
+    }
+    std::string name;
+    char state;
+    pid_t pid;
+    pid_t ppid;
+    uid_t euid;
+    uid_t ruid;
+    gid_t rgid;
+    int32_t priority = 0;
+    int32_t nice = 0;
+    uint64_t virtualMemorySizeKiB = 0;
+    __darwin_time_t startTime = 0;
```

Comment:
Default values on struct declaration? TIL this is valid since C++11 :smile:

Reply:
Yeah. I will remove these though – they are not necessary.